crypto/storage/azure/keyvault.go

```diff
@@ -0,0 +1,276 @@
+}
+
+// StorageType is the name of this storage type, used in health check reports and configuration.
+const StorageType = "azure-keyvault"
+
+type keyvault struct {
+	client  keyVaultClient
+	timeOut time.Duration
+	useHSM  bool
+}
+
+func (a keyvault) Name() string {
+	return StorageType
+}
+
+func (a keyvault) CheckHealth() map[string]core.Health {
+	return nil
+}
+
+func (a keyvault) NewPrivateKey(ctx context.Context, namingFunc func(crypto.PublicKey) (string, error)) (crypto.PublicKey, string, error) {
+	keyID, err := namingFunc(nil)
+	if err != nil {
+		return nil, "", err
+	}
+	// Make sure it doesn't already exist: Azure Key Vault otherwise creates a new version for the same key.
+	exists, err := a.PrivateKeyExists(ctx, keyID)
+	if err != nil {
+		return nil, "", err
+	}
+	if exists {
+		return nil, "", spi.ErrKeyAlreadyExists
+	}
+
+	keyName := keyIDToKeyName(keyID)
+
+	var keyType azkeys.KeyType
+	if a.useHSM {
+		keyType = azkeys.KeyTypeECHSM
+	} else {
+		keyType = azkeys.KeyTypeEC
+	}
+
+	response, err := a.client.CreateKey(ctx, keyName, azkeys.CreateKeyParameters{
+		Kty:   to.Ptr(keyType),
+		Curve: to.Ptr(azkeys.CurveNameP256),
+		KeyAttributes: &azkeys.KeyAttributes{
+			Enabled:    to.Ptr(true),
+			Exportable: to.Ptr(false),
+		},
+	}, nil)
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to create key in Azure Key Vault (name=%s): %w", keyName, err)
+	}
+	publicKey, _, err := parseKey(response.Key)
+	if err != nil {
+		return nil, "", err
+	}
+	return publicKey, keyID, nil
+}
+
+func (a keyvault) GetPrivateKey(ctx context.Context, kid string) (crypto.Signer, error) {
+	keyName := keyIDToKeyName(kid)
+	response, err := a.getPrivateKey(ctx, keyName)
+	if err != nil {
+		return nil, err
+	}
+	publicKey, signingAlgorithm, err := parseKey(response.Key)
+	if err != nil {
+		return nil, err
+	}
+	return &azureSigningKey{
+		client:           a.client,
+		timeOut:          a.timeOut,
+		keyName:          keyName,
+		publicKey:        publicKey,
+		signingAlgorithm: signingAlgorithm,
+	}, nil
+}
+
+func (a keyvault) PrivateKeyExists(ctx context.Context, kid string) (bool, error) {
+	_, err := a.getPrivateKey(ctx, keyIDToKeyName(kid))
+	if errors.Is(err, spi.ErrNotFound) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (a keyvault) DeletePrivateKey(ctx context.Context, kid string) error {
+	_, err := a.client.DeleteKey(ctx, keyIDToKeyName(kid), nil)
+	responseError := new(azcore.ResponseError)
+	if errors.As(err, &responseError) && responseError.StatusCode == http.StatusNotFound {
+		return spi.ErrNotFound
+	} else if err != nil {
+		return fmt.Errorf("unable to delete key from Azure Key Vault (name=%s): %w", keyIDToKeyName(kid), err)
+	}
+	return nil
+}
+
+func (a keyvault) getPrivateKey(ctx context.Context, keyName string) (*azkeys.GetKeyResponse, error) {
+	response, err := a.client.GetKey(ctx, keyName, "", nil)
+	responseError := new(azcore.ResponseError)
```

Comment:
how much overhead does it add to check this every time before using a key? If this succeeds once it probably will every time after that. Should the results be cached in `keyvault` and let Sign also check for 404 incase it has been deleted?

Reply:
Premature optimization, could always be added later (and it should at least be faster than the external storage API that requires 2 hops).